File: weektwo/revise/queandstack/Stack.java
package revise.queandstack;

import repeat.StackExceptions;

public class Stack {
    class Node {
        Node next;
        int data;

        Node(int data) {
            this.data = data;
        }
    }

    Node top;

    Stack() {
        this.top = null;
    }

    boolean isEmpty() {
        return top == null;
    }

    void push(int x) {
        Node temp = new Node(x);
        temp.next = top;
        top = temp;
    }

    int pop() throws StackExceptions {
        if (isEmpty()) {
            throw new StackExceptions("stack is empty");
        }

        int val = top.data;
        top = top.next;
        return val;
    }

    void display() {
        Node temp = top;
        while(temp!=null){
            System.out.print(temp.data+" --> ");
            temp = temp.next;
        }
        System.out.println();
    }

    public static void main(String[] args) throws StackExceptions {
        Stack s =new Stack();
        s.push(23);
        s.push(6);
        s.push(3);
        s.push(2);
        s.display();
        s.pop();
        s.display();
        s.pop();
        s.display();
        s.push(29);
        s.display();
    
    }
}

File: dsaJavascript/workout/extra/stack.js
class Stack {
  #items;
  constructor() {
    this.#items = [];
  }

  getStack() {
    return this.#items;
  }

  push(value) {
    this.#items.push(value);
  }

  pop() {
    if (!this.isEmpty()) {
      return this.#items.pop();
    }
    return null;
  }

  peek() {
    if (!this.isEmpty()) {
      return this.#items[this.#items.length - 1];
    }
    return null;
  }

  isEmpty() {
    return this.#items.length === 0;
  }
}

let stack = new Stack();

stack.push(23);
stack.push(1);
stack.push(8);
stack.push(3);
stack.push(7);

console.log(stack.peek());

console.log(stack.getStack());
